File: 152. Maximum Product Subarray/solution.py
# going by the java approach
class Solution:
    def maxProduct(self, nums: List[int]) -> int:

            # result variable
            r= nums[0]
            imin = nums[0]
            imax = nums[0]

            # lets loop
            for i in range(1,len(nums)):

                # swap if negative
                if nums[i] < 0 :

                    #swap manually
                    temp = imin
                    imin = imax
                    imax = temp

                imax = max(nums[i],imax*nums[i])
                imin = min(nums[i],imin*nums[i])
                r= max(r,imax)

            # return r
            return r

File: 152. Maximum Product Subarray/solution.cpp
// using the same approach as in the java program
class Solution {
public:
    int maxProduct(vector<int>& nums) {
      int r = nums[0];
      for (int i = 1, imin = r, imax = r; i < nums.size(); i++) {
              if (nums[i]<0) {
		// interestingly swap is also a function in c++
		// reference : https://www.geeksforgeeks.org/swap-in-cpp/
                swap(imin,imax);
              }
              imax = max(nums[i],imax*nums[i]);
              imin = min(nums[i],imin*nums[i]);
              r = max(r,imax);
      }
      return r;
    }
};
